src/java.base/share/classes/sun/security/jca/JCAUtil.java

```diff
@@ -69,4 +70,27 @@ public static SecureRandom getSecureRandom() {
         return CachedSecureRandomHolder.instance;
     }
 
+    // called by sun.security.jca.Providers class when provider list is changed
+    static void clearDefSecureRandom() {
+        def = null;
+    }
+
+    /**
+     * Get the default SecureRandom instance. This method is the
+     * optimized version of "new SecureRandom()" which re-uses the default
+     * SecureRandom impl if the provider table is the same.
+     */
+    public static SecureRandom getDefSecureRandom() {
+        SecureRandom result = def;
+        if (result == null) {
+            synchronized (JCAUtil.class) {
```

Comment:
Could this lock be avoided if set the value in the Providers update (or when the providers list is updated)?

Reply:
Well, then we have to pay the cost of "new SecureRandom()" at every provider list update when it may not be needed. It's hard to say which way is better... I thought about it and it seems more reasonable to only pay the cost when the SecureRandom object is needed.

Reply:
It makes sense to me.